### PR TITLE
Lazy-ily compute sidecar scope for namespaces without Sidecar

### PR DESCRIPTION
**Pull request by howardjohn**
Fixes https://github.com/istio/istio/issues/32034

This is both an optimization and fix for ^.

Before, we iterated over all namespaces that had services and computed
sidecar scopes for them. This is inaccurate and inefficient: a namespace
may have 0 services (inaccurate), and it may have services but no
proxies (inefficient).

This changes it such that on each request, we lazy load the sidecar
scope for that namespace.

NOTE: this represents a fairly substantial change to push context in
that its not as much of a "snapshot" anymore. It still is deterministic
what the output is given a push version (ie computing this up front, or
lazy loaded will return the same result), but the actual PushContext
object is mutated. While I don't love this, I think its ok because its
an optimization matter (ie we could remove it entirely, it would just
make things super slow).

### CODE REVIEW

File: pilot/pkg/model/push_context.go [hunk 1]
```diff
@@ -130,17 +130,24 @@ type sidecarIndex struct {
...
 	// the Sidecar for the root namespace (if present). This applies to any namespace without its own Sidecar.
 	rootConfig *config.Config
-	// defaultSidecarsByNamespace contains the default sidecar for namespaces that do not have a sidecar.
+	// computedSidecarsByNamespace contains the default sidecar for namespaces that do not have a sidecar.
+	// These may be DefaultSidecarScopeForNamespace if rootConfig is empty or ConvertToSidecarScope if not.
 	// These are lazy-loaded. Access protected by defaultSidecarMu
-	defaultSidecarsByNamespace map[string]*SidecarScope
-	defaultSidecarMu           *sync.Mutex
+	computedSidecarsByNamespace map[string]*SidecarScope
+	// computedSidecarsByNamespace contains the default sidecar for namespaces that do not have a sidecar,
```
Comment (ramaraochavali) [inline pilot/pkg/model/push_context.go:137]: nit computedSidecarsByNamespace -> gatewayDefaultSidecarsByNamespace

File: pilot/pkg/model/push_context.go [hunk 2]
```diff
@@ -820,7 +847,36 @@ func (ps *PushContext) getSidecarScope(proxy *Proxy, workloadLabels labels.Colle
 		}
 	}
 
-	return DefaultSidecarScopeForNamespace(ps, proxy.ConfigNamespace)
+	// We didn't have a Sidecar in the namespace. This means we should use the default - either an implicit
+	// default selecting everything, or pulling from the namespace.
```
Comment (ramaraochavali) [inline pilot/pkg/model/push_context.go:851]: nit: pulling from the namespace. -> pulling from the root namespace?

File: pilot/pkg/model/push_context.go [hunk 3]
```diff
@@ -820,7 +847,36 @@ func (ps *PushContext) getSidecarScope(proxy *Proxy, workloadLabels labels.Colle
...
+		// We have already computed the scope for this namespace, just fetch it
+		return sc
+	}
+	// We need to compute this namespace
+	var computed *SidecarScope
+	if ps.sidecarIndex.rootConfig != nil {
+		computed = ConvertToSidecarScope(ps, ps.sidecarIndex.rootConfig, proxy.ConfigNamespace)
+	} else {
+		computed = DefaultSidecarScopeForNamespace(ps, proxy.ConfigNamespace)
+		// Even though we were a gateway, we can store this as a gateway one since it could be used by a gateway
```
Comment (ramaraochavali) [inline pilot/pkg/model/push_context.go:875]: nit:  Even though we were a gateway ->  Even though this is for sidecar?

**Review by ramaraochavali — COMMENTED** [review]
some minor comment nits. 

And one question - if we do it lazily - do we have to maintain this in push context? Should we just do it when proxy is initialized ?

**Comment by howardjohn** [comment]
> some minor comment nits.
> 
> And one question - if we do it lazily - do we have to maintain this in push context? Should we just do it when proxy is initialized ?

It's basically an optimization. Instead of computing 1000 times for 1000 proxies in namespace we just do it once and share the result

**Comment by ramaraochavali** [comment]
> It's basically an optimization. Instead of computing 1000 times for 1000 proxies in namespace we just do it once and share the result

nvm.. I understood this part was thinking about some thing else .

**Review by ramaraochavali — APPROVED** [review]
LGTM. Put a hold to address the minor comments. Feel free to remove it once they are done.

File: pilot/pkg/model/push_context.go [hunk 4]
```diff
@@ -820,7 +847,36 @@ func (ps *PushContext) getSidecarScope(proxy *Proxy, workloadLabels labels.Colle
...
+	if f {
+		// We have already computed the scope for this namespace, just fetch it
+		return sc
+	}
+	// We need to compute this namespace
+	var computed *SidecarScope
+	if ps.sidecarIndex.rootConfig != nil {
+		computed = ConvertToSidecarScope(ps, ps.sidecarIndex.rootConfig, proxy.ConfigNamespace)
+	} else {
+		computed = DefaultSidecarScopeForNamespace(ps, proxy.ConfigNamespace)
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/push_context.go:874]: isn't  this duplicate with L860

Comment (howardjohn) [reply pilot/pkg/model/push_context.go:874]: the code is duplicated but only this or L860 will run, never both

Comment (hzxuzhonghu) [reply pilot/pkg/model/push_context.go:874]: Correct, it confuses me with the below line.  We may not have gateways reside in same namespace as normal sidecar proxy

**Pull request merged**